Image wrapper image class

chrisc2244 opened the pull request:
Adds functionality for ImageWrapper and a basic Image class.

We can use the decodeImageToRGBAVector() method to generate raw RGBA data from an image in the form of a vector, then use convertRGBAVectorToPixelVector() to generate the final vector<pixel> we need to start making Image objects. 

Also, I went with 'lodepng' to process the PNGs instead of 'spng' because it was way easier to use, that's why lodepng.cpp and lodepng.h are included in this. I tried to add comments within the code to also try and explain how everything worked.

Definitely take a look at the methods that deal with the massive vectors, iirc the decodeImageToRGBAVector() returns a vector that's like 4 million elements long, convertRGBAToPixelVector() returns a vector that's like 1 million pixel objects long, and I still suck at pointers + references so make sure I'm not accidentally copying them over and over or something.

chrisc2244 committed:
Adds some ImageWrapper functionality and a basic Image class.

chrisc2244 committed:
fixed the main function, I had changed it to do some testing

Jon-Cavaliere commented on ImageWrapper.cpp:
```diff
@@ -0,0 +1,70 @@
+#include <vector>
+#include <iostream>
+#include "ImageWrapper.h"
+//#include "Log.h" idk why but this isnt working
```
The path to this file is src/utils/Log.h

So #include "utils/Log.h" should work here

Jon-Cavaliere replied:
Actually you didn't put these files in the src directory, you put them in the root folder,
so you'd have to #include "src/utils/Log.h" or move the new files into src

Jon-Cavaliere commented on Image.h:
```diff
@@ -0,0 +1,35 @@
+
+	std::string m_filename;
+	std::vector<pixel> m_decodedpixels;
+	//std::string m_correctDiagnosis; 
```
Lets put this in the Patient class, but perhap make it a vector of strings because some have multiple diagnoses

Jon-Cavaliere commented on ImageWrapper.cpp:
```diff
@@ -0,0 +1,70 @@
+
+	std::vector<unsigned char> imagepixels;
+	 //the raw pixels, size: 8 bits, range: 0 -> 255
+	unsigned int width, height;
```
Better practice is to use the datatype uint32_t because it's always the same across platforms and compilers where unsigned int may vary

Jon-Cavaliere commented on ImageWrapper.cpp:
```diff
@@ -0,0 +1,70 @@
+	unsigned error = lodepng::decode(imagepixels, width, height, filename);
+
+	//if there's an error, display it
+	if (error) std::cout << "decoder error " << error << ": " << lodepng_error_text(error) << std::endl;
```
MIR::Log::writeEr("ImageWrapper->decodeImageToRGBAVector()", lodepng_error_text(error);

Jon-Cavaliere commented on ImageWrapper.h:
```diff
@@ -0,0 +1,42 @@
+
+struct pixel
+{
+	int r; //0 -> 255 red
```
uint8_t r, g, b; // C++11 standard datatype for a 1 byte number

same as unsigned char (this is the old option)

Has a range from 0-255 so it would be more memory efficient.

Jon-Cavaliere replied:
Apparently C++ 17 has std::byte defined, but our project is using the C++14 standard

Jon-Cavaliere commented on ImageWrapper.cpp:
```diff
@@ -0,0 +1,70 @@
+//increments by 4 through RGBAVector to fill out pixel struct, then adds to pixelVectorToReturn then returns it.
+//maybe some optimization here somehow through pointers/refs, i should definitely not pass
+//an entire copy of RGBAVector
+std::vector<pixel> ImageWrapper::convertRGBAVectorToPixelVector(std::vector<unsigned char>& RGBAVector)
```
You're already passing the vector as a reference! perhaps just making it a const reference here is a good little optimization for now since we aren't changing / mutating the vector at all in this method

std::vector<pixel> ImageWrapper::convertRGBAVectorToPixelVector(const std::vector<unsigned char>& RGBAVector)

Jon-Cavaliere commented on ImageWrapper.cpp:
```diff
@@ -0,0 +1,70 @@
+		pixelVectorToReturn.push_back(p);
+	}
+
+	return pixelVectorToReturn;
```
This should fine, in older c++ it would copy the whole local vector you created here when you return it, but since C++11, std::vector has move-semantics, meaning the local vector you create here should be moved upon return.

Jon-Cavaliere replied:
And I believe that you have this project set to use C++14 standards

Jon-Cavaliere commented on Image.h:
```diff
@@ -0,0 +1,35 @@
+
+private:
+
+	std::string m_filename;
```
Store image width and height here too!

uint32_t m_width, m_height

Jon-Cavaliere replied:
Also adding some accessor methods like 

const uint32_t getWidth() const;
Etc..

Jon-Cavaliere reviewed (changes requested):
Looks good so far! I've suggested a few changes with the dual purpose of optimizing your code and teaching you a little bit more about modern C++.

chrisc2244 committed:
-Added width/height parameters to Image class, along with accessor methods
-moved pixel vector to Image object, put Image object as parameter of Patient class
-added logging to ImageWrapper decode method
-converted unsigned char -> uint8_t (except in decode method)
-converted unsigned int / int-> uint16_t (except in decode method)
-convertRGBAVector method in ImageWrapper to take const reference
-(hopefully?) converted project to c++20, followed a guide on google, dunno if it worked for sure, if it did we could use std::byte

chrisc2244 committed:
hadn't actually saved the files before I pushed the last commit. Here is the actual one.

Jon-Cavaliere committed:
Various fixes;

Loading the complete dataset and populating patient vector is done

added Image constructor that just takes a filename, but will not load the
png until we need to.

Jon-Cavaliere committed:
Moved 3rd Party Lib -> Vendor

The vendor folder in src will contain any third party libraries that we use

Jon-Cavaliere committed:
Refactored CSV

Fixed the warnings CSV_Util and Table
Fixed Copy Paste Error xD

Jon-Cavaliere committed:
Added Header Comments

Added some descriptive comments to CSV classes

Jon-Cavaliere committed:
exit failure method

call exitFailure() when a fatal exception is encountered in order to properly
close the log for debugging and review later

See Application::init() for example use

Jon-Cavaliere committed:
Application Singleton

Application class is a singleton so we can call
Application::getInstance()->exitFailure() anywhere in the code

Jon-Cavaliere committed:
Add comment; log

Jon-Cavaliere committed:
Renamed getWidth() in CSV_Table

getWidth() is now get sizeRows()
getHeight() is now sizeColumns()

it just makes more sense
